presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -47,6 +47,8 @@
     private final Optional<Decompressor> decompressor;
     private final Optional<SpillCipher> spillCipher;
 
+    private byte[] compressionBuffer;
```

Comment:
My worry about this approach is basically the same as [I mentioned on a previous PR](https://github.com/prestodb/presto/pull/13032#discussion_r302621008): this buffer will grow to `compressor.maxCompressedLength(uncompressedSize)` for the largest input passed through and will not be elligible for garbage collection until the whole `PagesSerde` instance is. This will be true for all `PagesSerde` instances created, so while it will indeed reduce GC pressure (and time spent zeroing out new byte[] allocations) it could risk increasing GC pressure and memory usage when heap is constrained. This is less of a concern if all pages serialized are small, but serializing one large page will having lasting effects on the memory footprint of a `PagesSerde` with this implementation. Furthermore, this memory usage is untracked and therefore invisible to memory budgets and bookkeeping.

I see a few options to mitigate this concern:
- Restrict the maximum size of serialized pages to set an upper bound on memory pressure
- Restrict the maximum size of the `compressionBuffer`. If maxCompressedLength is greater than this limit, a new byte array is created and not stored for future use.
- Add a method that allows callers to signal that the buffer should be released. This will require code changes for all usages of `PagesSerde` but would make it easier to account for query memory tracking.
- Pass a shared buffer pool into the `PagesSerde#serialize` method. This would allow buffer reuse _between_ instances and decouple the lifespan of the buffer from the lifespan of the `PagesSerde`.

@mbasmanova thoughts on the options mentioned above?

Reply:
This is a legit point. "Restrict the maximum size" sounds a reasonable approach. Can we profile the distribution of `maxCompressedLength` in production? This can give us the insight to decide the max size we may need.

Reply:
@pettyjamesm @highker I'm seeing that we limit the size of the pages being serialized to 1MB. Perhaps, we could add a `checkArgument(page.getSizeInBytes() <= DEFAULT_MAX_PAGE_SIZE_IN_BYTES)` to `serialize` method. What do you think?

```
public static final int DEFAULT_MAX_PAGE_SIZE_IN_BYTES = 1024 * 1024;

        List<SerializedPage> serializedPages = splitPage(page, DEFAULT_MAX_PAGE_SIZE_IN_BYTES).stream()
                .map(serde::serialize)
                .collect(toImmutableList());
```

Reply:
Hi all, thanks for your inputs. 

Currently PagesSerde objects are used by PartitionedOutputOperator, TaskOutputOperator, ExchangeOperator, MergeOperator and Query class. The first two operators deal with the serialization while the last 3 deals with deserialization and not in the scope of this PR. By looking at the usse cases of PartitionedOutputOperator and TaskOutputOperator, I don't think keeping the compression buffer as a member of PagesSerde a big problem: 

1) There is only one PagesSerde object per operator, not per table column or per destination partition. And there's only one compressionBuffer per PagesSerde.

2) The sizes of the SerializedPage being compressed would be pretty much uniform, closing to DEFAULT_MAX_PAGE_SIZE_IN_BYTES (1MB by default). For the new OptimizedPartitionedOutputOperator this size is almost always close to DEFAULT_MAX_PAGE_SIZE_IN_BYTES. The old PartitionedOutputOperator has slightly higher variance but the size is still quite predictable. This is because we always accumulate data up to this size before the serialization and flushing happens, except for the last page. And the compressed size should be less than or equal to the uncompressed size, which is bounded by 1MB almost all time. There are only a fixed number of threads on each worker node and the overhead would be at the level of hundreds of MB per node.

3) The PagesSerde objects only lives through these operators but not the global life time of Presto process. The PagesSerde objects and the compression buffer will be collected after the lifetime of these operators (stage execution) ends. If we don't keep the compression buffer as a member, the operators would nonetheless keep trying to allocate this 1MB buffer for every SerializedPage over and over again throughout the lifetime of the operator. while the proposed change only maintains this 1MB buffer during the same period of time. Sure the compression buffer becomes long lived memory and can't be GCed after the compression is done and before the operator dies, but overhead is small and predictable as we discussed in previous 2 points. I would expect the thousands or even millions of allocation of this buffer contribute to OOM problem more than what materializing this buffer does. 

I agree we should track the memory in systemMemoryContext. I will make the change in next iteration. The proposed way is to always count the max size possible, which is the maxCompressedLength(DEFAULT_MAX_PAGE_SIZE_IN_BYTES).

Reply:
With the limitation on size and the addition of adding memory tracking, I’m no longer especially concerned with the approach.

I’d rather see a buffer reuse abstraction passed, but that can come later. I’m still in favor of that strategy for a couple reasons:
- it can easily extend the benefit to encryption as well as compression (assuming the encryption implementation in prestosql is ported back over)
- it can reduce the total allocation rate by sharing and recycling buffers between instances as well as centrally manage amount of memory used
- some deployments may be less interested in the performance benefit and more averse to the (admittedly low in this case) risk of cross-query data access. Those deployments could opt to use a “never reuse” pool.

Reply:
Hi @pettyjamesm By "buffer reuse abstraction passed", did you mean pass a shared buffer pool into the PagesSerde#serialize method? We have some work on the global array pool, and hopefully we'll have time to pick it up this half. For this PR, I can create a config property "experimental.repartition_reuse_compression_buffer" and default it to true. 

However for this PR, there was no cross-query data access by reusing the compression buffer. The buffer is owned by the PagesSerde object, which is in term owned by the operator, and the operator is created per query and not reused by many queries.

Reply:
Yeah, that’s correct. I don’t think adding the config flag is necessary in this PR since the PagesSerDe instance has an implicit per query lifespan already, but it seems obvious that the caller has more context about the lifespan of buffers than embedding the buffer as a field as you’ve done here. For instance, the shared compression buffer needn’t make a copy out of the shared buffer in the case of say, spill to disk since the spiller knows that it can be released after being written to disk but the PagesSerDe doesn’t know that context.

Reply:
Thanks @pettyjamesm. With the global array pool things will make more sense but that requires much more careful work.  Back to this PR, do you think we need to do anything else?

Reply:
Nope, I’m satisfied with the current state of tracking the buffer memory and enforcing an upper bound on its size.